Automattic/wp-calypso
# Login: Add Content Security Policy header

### Background
[Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) (CSP) serves as a mitigation against cross-site scripting attacks (XSS).

[More information in this blog post](https://hackernoon.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5)
Additional resources: [Google](https://developers.google.com/web/fundamentals/security/csp/) and [OWASP](https://www.owasp.org/index.php/Content_Security_Policy_Cheat_Sheet)

Tool to evaluate policy: https://csp-evaluator.withgoogle.com/

This PR aims to introduce the following policy to `/log-in/*`:
```
default-src 'self'; script-src 'self' 'unsafe-eval' stats.wp.com https://apis.google.com 'nonce-bfae467b9aaca24fc9acc9b0491128220dc551c30f487d6d6fd04523023270a83880c6583d7526593e3cf943fdce4fc2' 'nonce-b7c041b5271937931404f54ae3df321ac775ea8ea1d6ad00e1f109fdc2daaa4026cd2043d7a23cebaca1a35f06f73625' 'sha256-3yiQswl88knA3EhjrG5tj5gmV6EUdLYFvn2dygc0xUQ=' 'sha256-ZKTuGaoyrLu2lwYpcyzib+xE4/2mCN8PKv31uXS3Eg4='; base-uri 'none'; style-src 'self' *.wp.com https://fonts.googleapis.com; form-action 'self'; object-src 'none'; img-src 'self' *.wp.com https://www.google-analytics.com; frame-src 'self' https://public-api.wordpress.com https://accounts.google.com/; font-src 'self' *.wp.com https://fonts.gstatic.com data:; media-src 'self'; connect-src 'self' https://wordpress.com/; report-uri /cspreport
```
in a report only mode, for now.

While it would be beneficiary to have this for all of calypso, IMO we should start with something small and contained, such as the `/log-in/*` pages.

#### Testing instructions
  
1. Run `git checkout add/csp` and start your server, or open a [live branch](https://calypso.live/log-in?branch=add/csp)
2. Open the [`Login` page](http://calypso.localhost:3000/log-in)
3. Assert that you indeed see the `Content-Security-Policy-Report-Only` header
  
#### Reviews
  


### TODO ( in following PRs ):
- [ ] `store` library uses eval to get global object, can be solved by version upgrade, but for now it violates `'unsafe-eval'`
- [ ] `is-my-json-valid` depends on `generate-function` which violates `'unsafe-eval'` ( I don't see an obvious solution to this )
- [x]  Add hash/nonce to google analytics library to avoid whitelisting the host
